### PR TITLE
Refactor ZStream.fromQueue

### LINKED ISSUE #1380: ZStream#fromQueue interruption
Currently ZStream#fromQueue is implemented like this
```scala
  final def fromQueue[R, E, A](queue: ZQueue[_, _, R, E, _, A]): ZStream[R, E, A] =
    unfoldM(()) { _ =>
      queue.take
        .map(a => Some((a, ())))
        .foldCauseM(
          cause =>
            // Dequeueing from a shutdown queue will result in interruption,
            // so use that to signal the stream's end.
            if (cause.interrupted) ZIO.succeed(None)
            else ZIO.halt(cause),
          ZIO.succeed
        )
    }
```
It hides the effect that upstream has been interrupted / the queue has been shutdown.
As most of the ZStream code uses Take to precisely pass around when / why a stream has ended I wonder if this is still useful behavior

### PR DESCRIPTION
**Pull request by regiskuckaertz**
Closes #1380 

It seems like a more efficient formulation too.

### CODE REVIEW

**Comment by regiskuckaertz** [comment]
> we keep the semantics of end-on-interrupt in `fromQueue`

do you agree that this is only when the queue has been shutdown? if the interruption happens because of some other cause, then it should be propagated? something like this then:
```scala
queue.take.catchAllCause { c =>
  queue.isShutdown.flatMap(if (_ && c.interrupted) Pull.end else Pull.halt(c))
}
```

**Comment by iravid** [comment]
Yes, sorry! Meant to write "end-on-shutdown".

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 1]
```diff
@@ -2125,17 +2125,21 @@ object ZStream extends ZStreamPlatformSpecific {
...
+          UIO(
+            queue.take.catchAllCause(
+              c => queue.isShutdown.flatMap(down => if (down && c.interrupted) Pull.end else Pull.halt(c))
+            )
+          ),
+          _ => queue.shutdown
         )
+      )
+    }
+          _ => queue.shutdown
```
Comment (regiskuckaertz) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: 🤔 that's not at all what I just pushed 🤣

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 2]
```diff
@@ -2125,17 +2125,34 @@ object ZStream extends ZStreamPlatformSpecific {
...
+          ),
+          _ => queue.shutdown
         )
+      )
+    }
+
+  /**
+   * Creates a stream from a [[zio.ZQueue]] of queue `take` operations
+   */
+  final def fromQueueTake[R, E, A](queue: ZQueue[_, _, R, E, _, Take[E, A]]): ZStream[R, E, A] =
```
Comment (regiskuckaertz) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: @iravid Did I understand your suggestion correctly? ☝️

Comment (iravid) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: Yup! Although, I realize now that this is just `fromQueue(q).unTake`. Do you think this justifies an additional constructor?

File: streams/shared/src/main/scala/zio/stream/ZStream.scala [hunk 3]
```diff
@@ -2125,17 +2125,34 @@ object ZStream extends ZStreamPlatformSpecific {
...
-          ZIO.succeed
+    ZStream[R, E, A] {
+      ZManaged.reserve(
+        Reservation(
+          UIO(
+            queue.take.catchAllCause(
+              c => queue.isShutdown.flatMap(down => if (down && c.interrupted) Pull.end else Pull.halt(c))
+            )
+          ),
+          _ => queue.shutdown
```
Comment (iravid) [inline streams/shared/src/main/scala/zio/stream/ZStream.scala]: I also suggest not shutting down the queue - there's `fromQueueWithShutdown` for that :-)

Comment (regiskuckaertz) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: Oh, I didn't see that. Fair enough 👍

File: streams/shared/src/main/scala/zio/stream/ZStream.scala (hunk 2 above)
Comment (regiskuckaertz) [reply streams/shared/src/main/scala/zio/stream/ZStream.scala]: Nope! ok, undoing that one.

**Comment by iravid** [comment]
Thanks @regiskuckaertz!

**Pull request merged**